File: LeetCode/LC.3113.find-the-number-of-subarrays-where-boundary-elements-are-maximum.cpp
#include "../utils/abel_macro.h"

class Solution {
public:
    long long numberOfSubarrays(vector<int>& nums) {
        long long ans = nums.size();    // single-element subarray
        vector<pair<int,int>> st;   // mono stack: (num, cnt)
        st.emplace_back(INT_MAX, 1);
        for (int x : nums) {
            while (st.back().first < x) {
                st.pop_back();
            }
            if (st.back().first == x) {
                ans += st.back().second;
                st.back().second++;
            } else {
                st.emplace_back(x, 1);
            }
        }
        return ans;
    }
};

File: LeetCode/LC.3113.find-the-number-of-subarrays-where-boundary-elements-are-maximum.py
from lc_utils import *


class Solution:
    def numberOfSubarrays(self, nums: List[int]) -> int:
        ans = len(nums)
        st = [[inf, 0]]  # sentinel
        for x in nums:
            while x > st[-1][0]:
                st.pop()
            if x == st[-1][0]:
                ans += st[-1][1]
                st[-1][1] += 1
            else:
                st.append([x, 1])
        return ans
